mistralrs-core/src/sampler.rs

```diff
@@ -247,22 +247,22 @@ impl Sampler {
     }
 
     fn apply_repeat_presence_penalty(
-        logits: &Tensor,
+        vocab_size: usize,
         presence_penalty: f32,
         frequency_penalty: f32,
         context: &[u32],
-    ) -> Result<Tensor> {
+    ) -> Result<Vec<f32>> {
         //mu[j] -> mu[j] - c[j] * alpha_frequency - float(c[j] > 0) * alpha_presence
-        let device = logits.device();
-        let mut logits = logits.to_vec1::<f32>()?;
+        // let device = logits.device();
+        // let mut logits = logits.to_vec1::<f32>()?;
+        let mut logits = vec![0.0; vocab_size];
         for (token_id, logit) in logits.iter_mut().enumerate() {
             let count = context.iter().filter(|x| **x as usize == token_id).count();
             *logit = *logit
                 - count as f32 * frequency_penalty
                 - if count > 0 { 1. } else { 0. } * presence_penalty;
         }
-        let logits_len = logits.len();
-        Tensor::from_vec(logits, logits_len, device)
+        Ok(logits)
```

Comment:
Since this reads the context we can't pre-allocate the tensor.

Allocating the tensor here was triggering a synchronization that would cost 500ms